File: app/src/main/java/com/boardgamegeek/filterer/YearPublishedFilterer.java
package com.boardgamegeek.filterer;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.util.MathUtils;
import com.boardgamegeek.util.StringUtils;

import java.util.Calendar;
import java.util.Locale;

public class YearPublishedFilterer extends CollectionFilterer {
	public static final int MIN_RANGE = 1970;
	public static final int MAX_RANGE = Calendar.getInstance().get(Calendar.YEAR) + 1;

	private int min;
	private int max;

	public YearPublishedFilterer(Context context) {
		super(context);
	}

	public YearPublishedFilterer(Context context, int min, int max) {
		super(context);
		this.min = min;
		this.max = max;
	}

	@Override
	public void setData(@NonNull String data) {
		String[] d = data.split(DELIMITER);
		min = d.length > 0 ? MathUtils.constrain(StringUtils.parseInt(d[0], MIN_RANGE), MIN_RANGE, MAX_RANGE) : MIN_RANGE;
		max = d.length > 1 ? MathUtils.constrain(StringUtils.parseInt(d[1], MAX_RANGE), MIN_RANGE, MAX_RANGE) : MAX_RANGE;
	}

	@Override
	public int getTypeResourceId() {
		return R.string.collection_filter_type_year_published;
	}

	@Override
	public String getDisplayText() {
		String text;
		String minText = String.valueOf(min);
		String maxText = String.valueOf(max);

		if (min == MIN_RANGE && max == MAX_RANGE) {
			text = "";
		} else if (min == MIN_RANGE) {
			text = maxText + "-";
		} else if (max == MAX_RANGE) {
			text = minText + "+";
		} else if (min == max) {
			text = maxText;
		} else {
			text = minText + "-" + maxText;
		}

		return text;
	}

	@Override
	public String getDescription() {
		@StringRes int prepositionResId;
		String year;
		if (min == MIN_RANGE && max == MAX_RANGE) {
			return "";
		} else if (min == MIN_RANGE) {
			prepositionResId = R.string.before;
			year = String.valueOf(max + 1);
		} else if (max == MAX_RANGE) {
			prepositionResId = R.string.after;
			year = String.valueOf(min - 1);
		} else if (min == max) {
			prepositionResId = R.string.in;
			year = String.valueOf(max);
		} else {
			prepositionResId = R.string.in;
			year = String.valueOf(min) + "-" + String.valueOf(max);
		}
		return context.getString(R.string.published_prefix, context.getString(prepositionResId), year);
	}

	@Override
	public String getSelection() {
		String format;
		if (min == MIN_RANGE && max == MAX_RANGE) {
			format = "";
		} else if (min == MIN_RANGE) {
			format = "%1$s<=?";
		} else if (max == MAX_RANGE) {
			format = "%1$s>=?";
		} else if (min == max) {
			format = "%1$s=?";
		} else {
			format = "(%1$s>=? AND %1$s<=?)";
		}
		return String.format(Locale.getDefault(), format, Games.YEAR_PUBLISHED);
	}

	@Override
	public String[] getSelectionArgs() {
		if (min == MIN_RANGE && max == MAX_RANGE) {
			return null;
		} else if (min == MIN_RANGE) {
			return new String[] { String.valueOf(max) };
		} else if (max == MAX_RANGE) {
			return new String[] { String.valueOf(min) };
		} else if (min == max) {
			return new String[] { String.valueOf(min) };
		} else {
			return new String[] { String.valueOf(min), String.valueOf(max) };
		}
	}

	public int getMin() {
		return min;
	}

	public int getMax() {
		return max;
	}

	@NonNull
	@Override
	public String flatten() {
		return String.valueOf(min) + DELIMITER + String.valueOf(max);
	}
}

File: app/src/main/java/com/boardgamegeek/filterer/YearPublishedFilterer.kt
package com.boardgamegeek.filterer

import android.content.Context
import androidx.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.model.CollectionItem
import com.boardgamegeek.extensions.IntervalDelegate
import com.boardgamegeek.extensions.andLess
import com.boardgamegeek.extensions.andMore
import java.util.*

class YearPublishedFilterer(context: Context) : CollectionFilterer(context) {
    var min by IntervalDelegate(LOWER_BOUND, LOWER_BOUND, upperBound)
    var max by IntervalDelegate(upperBound, LOWER_BOUND, upperBound)

    override val typeResourceId = R.string.collection_filter_type_year_published

    override val iconResourceId: Int
        get() = R.drawable.ic_baseline_calendar_today_24

    override fun inflate(data: String) {
        data.split(DELIMITER).run {
            min = getOrNull(0)?.toIntOrNull() ?: LOWER_BOUND
            max = getOrNull(1)?.toIntOrNull() ?: upperBound
        }
    }

    override fun deflate() = "$min$DELIMITER$max"

    override fun chipText(): String {
        val year = describeRange("-")
        return when {
            min == LOWER_BOUND && max == upperBound -> return ""
            min == LOWER_BOUND -> year.andLess()
            max == upperBound -> year.andMore()
            else -> year
        }
    }

    override fun description(): String {
        val year: String = describeRange()
        @StringRes val prepositionResId: Int = when {
            min == LOWER_BOUND && max == upperBound -> return ""
            min == LOWER_BOUND -> R.string.before
            max == upperBound -> R.string.after
            min == max -> R.string.`in`
            else -> R.string.`in`
        }
        return context.getString(R.string.published_prefix, context.getString(prepositionResId), year)
    }

    @Suppress("SameParameterValue")
    private fun describeRange(rangeSeparator: String = " - "): String {
        return when {
            min == LOWER_BOUND && max == upperBound -> return ""
            min == LOWER_BOUND -> max.toString()
            max == upperBound -> min.toString()
            min == max -> max.toString()
            else -> "$min$rangeSeparator$max"
        }
    }

    override fun filter(item: CollectionItem): Boolean {
        return when {
            min == LOWER_BOUND && max == upperBound -> true
            min == LOWER_BOUND -> item.collectionYearPublished <= max
            max == upperBound -> item.collectionYearPublished >= min
            min == max -> item.collectionYearPublished == min
            else -> item.collectionYearPublished in min..max
        }
    }

    companion object {
        const val LOWER_BOUND = 1970
        val upperBound = Calendar.getInstance().get(Calendar.YEAR) + 1
    }
}


